### PR TITLE
PERF: Datetime/Timestamp.normalize for timezone naive datetimes

### PR DESCRIPTION
**Pull request by mroeschke**
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

```
       before           after         ratio
     [bb32564c]       [ff171f75]
-         281±2μs          197±2μs     0.70  timeseries.DatetimeIndex.time_normalize('dst')
-     4.26±0.05ms      2.64±0.04ms     0.62  timeseries.DatetimeAccessor.time_dt_accessor_normalize
-     3.78±0.01ms      2.28±0.01ms     0.60  timeseries.DatetimeIndex.time_normalize('repeated')
-     3.79±0.05ms      2.23±0.06ms     0.59  timeseries.DatetimeIndex.time_normalize('tz_naive')


       before           after         ratio
     [bb32564c]       [ff171f75]
-      11.9±0.9μs         1.56±0μs     0.13  timestamp.TimestampOps.time_normalize(None)
```

### CODE REVIEW

File: pandas/_libs/tslibs/conversion.pyx [hunk 1]
```diff
@@ -1130,21 +1130,11 @@ def normalize_i8_timestamps(int64_t[:] stamps, object tz=None):
...
     cdef:
-        Py_ssize_t i, n = len(stamps)
-        npy_datetimestruct dts
+        Py_ssize_t n = len(stamps)
         int64_t[:] result = np.empty(n, dtype=np.int64)
 
-    if tz is not None:
-        tz = maybe_get_tz(tz)
-        result = _normalize_local(stamps, tz)
-    else:
```
Comment (jbrockmendel) [inline pandas/_libs/tslibs/conversion.pyx]: Is this case never reached?

Comment (mroeschke) [reply pandas/_libs/tslibs/conversion.pyx]: Correct. This case (the naive case) is handled in these two places now:

https://github.com/pandas-dev/pandas/pull/23634/files#diff-231ac35d2116a12844a7cfed02730580R1289
https://github.com/pandas-dev/pandas/pull/23634/files#diff-960da60b33d858481ad8799b0e84764bR835

File: pandas/_libs/tslibs/timestamps.pyx [hunk 2]
```diff
@@ -40,6 +40,7 @@ from timezones cimport (
 # Constants
 _zero_time = datetime_time(0, 0)
 _no_input = object()
+cdef int64_t DAY_NS = 86400000000000
```
Comment (jreback) [inline pandas/_libs/tslibs/timestamps.pyx]: we have DAY_NS defined in lots of places, can you move to 1

```
(bamboo-dev) jreback@dev:~/pandas-dev$ grep -r 86400 pandas/_libs/ --include '*.pyx'
pandas/_libs/tslibs/period.pyx:        {1, 24, 1440, 86400, 86400000, 86400000000, 86400000000000},
pandas/_libs/tslibs/period.pyx:        seconds = unix_date * 86400 + dts.hour * 3600 + dts.min * 60 + dts.sec
pandas/_libs/tslibs/period.pyx:        abstime += 86400
pandas/_libs/tslibs/period.pyx:    while abstime >= 86400:
pandas/_libs/tslibs/period.pyx:        abstime -= 86400
pandas/_libs/tslibs/period.pyx:    # abstime >= 0.0 and abstime <= 86400
pandas/_libs/tslibs/conversion.pyx:cdef int64_t DAY_NS = 86400000000000LL
pandas/_libs/tslibs/timedeltas.pyx:cdef int64_t DAY_NS = 86400000000000LL
pandas/_libs/tslibs/timedeltas.pyx:        m = 1000000000L * 86400 * 7
pandas/_libs/tslibs/timedeltas.pyx:        m = 1000000000L * 86400
pandas/_libs/tslibs/timedeltas.pyx:        86400000000042
pandas/_libs/tslibs/fields.pyx:    micros = np.mod(dtindex, 86400000000000, dtype=np.int64) // 1000LL
pandas/_libs/tslibs/src/datetime/np_datetime.c:    npy_int64 DAY_NS = 86400000000000LL;
```

prob should be in np_datetime.pyx (and import from there)

File: pandas/tests/scalar/timestamp/test_unary_ops.py [hunk 3]
```diff
@@ -328,6 +328,17 @@ def test_replace_dst_border(self):
         expected = Timestamp('2013-11-3 03:00:00', tz='America/Chicago')
         assert result == expected
 
+    # --------------------------------------------------------------
+    # Timestamp.normalize
+
+    @pytest.mark.parametrize('arg', ['2013-11-30', '2013-11-30 12:00:00'])
```
Comment (jreback) [inline pandas/tests/scalar/timestamp/test_unary_ops.py]: is there a normalize_nat test as well?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/scalar/timestamp/test_unary_ops.py (hunk 3 above)
Comment (mroeschke) [reply pandas/tests/scalar/timestamp/test_unary_ops.py]: We could have one for Timstamp mirroring (another issue). Probably would just return `NaT`

**Comment by mroeschke** [comment]
@jreback gathered some of the `DAY_NS` usage into a constant in np_datetime.pyx and all green.

File: pandas/_libs/tslibs/conversion.pyx [hunk 4]
```diff
@@ -22,7 +22,8 @@ from np_datetime cimport (check_dts_bounds,
                           npy_datetime,
                           dt64_to_dtstruct, dtstruct_to_dt64,
                           get_datetime64_unit, get_datetime64_value,
-                          pydatetime_to_dt64, NPY_DATETIMEUNIT, NPY_FR_ns)
+                          pydatetime_to_dt64, NPY_DATETIMEUNIT, NPY_FR_ns,
```
Comment (jreback) [inline pandas/_libs/tslibs/conversion.pyx]: what is DAY_S, you mean DAY_NS right? let's write out these constants.

File: pandas/_libs/tslibs/np_datetime.pyx [hunk 5]
```diff
@@ -37,6 +37,12 @@ cdef extern from "src/datetime/np_datetime_strings.h":
                                 npy_datetimestruct *out,
                                 int *out_local, int *out_tzoffset)
 
+# ----------------------------------------------------------------------
+# time constants
+
+cdef int64_t DAY_S = 86400
```
Comment (jreback) [inline pandas/_libs/tslibs/np_datetime.pyx]: let's write this out to DAY_SECONDS

File: pandas/_libs/tslibs/timestamps.pyx (hunk 2 above)
Comment (jreback) [reply pandas/_libs/tslibs/timestamps.pyx]: can you move to the same place you have DAY_SECONDS

File: pandas/core/arrays/datetimes.py [hunk 6]
```diff
@@ -832,7 +832,14 @@ def normalize(self):
                        '2014-08-01 00:00:00+05:30'],
                        dtype='datetime64[ns, Asia/Calcutta]', freq=None)
         """
-        new_values = conversion.normalize_i8_timestamps(self.asi8, self.tz)
+        if self.tz is None:
+            not_null = self.notnull()
```
Comment (jreback) [inline pandas/core/arrays/datetimes.py]: same

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/_libs/tslibs/np_datetime.pyx (hunk 5 above)
Comment (jbrockmendel) [reply pandas/_libs/tslibs/np_datetime.pyx]: I think the place for these may be ccalendar

File: pandas/core/arrays/datetimes.py (hunk 6 above)
Comment (jbrockmendel) [reply pandas/core/arrays/datetimes.py]: should this be `notna`?  (does `DatetimeArray` even _have_ notna or notnull?)

Comment (mroeschke) [reply pandas/core/arrays/datetimes.py]: It (`DatetimeIndex`) apparently has `notnull`, but not sure if i should be using `notna` or `notnull`

**Comment by mroeschke** [comment]
All green after some flaky azure tests.

**Review by jreback — APPROVED** [review]

File: pandas/_libs/tslibs/conversion.pyx [hunk 7]
```diff
@@ -41,7 +42,6 @@ from nattype cimport NPY_NAT, checknull_with_nat
 # ----------------------------------------------------------------------
 # Constants
 
-cdef int64_t DAY_NS = 86400000000000LL
 cdef int64_t HOURS_NS = 3600000000000
```
[the remark this answers is not in the document]
Comment (jreback) [reply pandas/_libs/tslibs/conversion.pyx]: can you move this one as well

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/arrays/datetimes.py (hunk 6 above)
Comment (jreback) [reply pandas/core/arrays/datetimes.py]: can you use notna

**Review by jreback — CHANGES_REQUESTED** [review]
tiny change. ping on green.

**Comment by mroeschke** [comment]
Ping all green @jreback

**Review by jreback — APPROVED** [review]

**Pull request merged**